pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,194 @@
+	}
+	return &newcontroller
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("podmetrics").With("pod", req.Name))
+
+	// Retrieve pod from reconcile request
+	pod := &v1.Pod{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to pod deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				podGaugeVec.Delete(*labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after pod labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		podGaugeVec.Delete(*labels)
+	}
+	newlabels, err := c.generateLabels(ctx, pod)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate new labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	gauge, err := podGaugeVec.GetMetricWith(*newlabels)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate new gauge: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	gauge.Set(float64(1))
+	c.LabelsMap[req.NamespacedName] = newlabels
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("podmetrics").
+		For(&v1.Pod{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10000}).
+		Complete(c)
+	return err
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(ctx context.Context, pod *v1.Pod) (*prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+	metricLabels[podName] = pod.GetName()
+	metricLabels[podNameSpace] = pod.GetNamespace()
+	// Selflink has been deprecated after v.1.20
+	// Manually generate the selflink for the first owner reference
+	// Currently we do not support multiple over references
+	selflink := ""
+	if len(pod.GetOwnerReferences()) != 0 {
+		ownerreference := pod.GetOwnerReferences()[0]
+		selflink = fmt.Sprintf("/apis/%s/namespaces/%s/%ss/%s", ownerreference.APIVersion, pod.Namespace, strings.ToLower(ownerreference.Kind), ownerreference.Name)
+	}
+	metricLabels[ownerSelfLink] = selflink
+	metricLabels[podHostName] = pod.Spec.NodeName
+	metricLabels[podPhase] = string(pod.Status.Phase)
+	provisioner := v1alpha5.DefaultProvisioner
+	if name, ok := pod.Spec.NodeSelector[v1alpha5.ProvisionerNameLabelKey]; ok {
+		provisioner.Name = name
+	}
+	metricLabels[podProvisioner] = provisioner.Name
+	nodename := types.NamespacedName{Name: pod.Spec.NodeName}
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, nodename, node); err != nil {
+		metricLabels[podHostZone] = "N/A"
```

Comment:
Do we need to include this as "N/A", or can it simply not exist or be ""

Reply:
Agree. Changing these to "".

Reply:
Turns out that Prometheus does not like "". It will basically create a null value for the label. It confused the concept of not having a label value and an empty value. I am using N/A so that it will be more readable to the users.